From 02dd014303d7a151398321cfc7001426306b6e3b Mon Sep 17 00:00:00 2001
From: suzuki toshiya <mpsuzuki@hiroshima-u.ac.jp>
Date: Wed, 26 Nov 2014 16:39:00 +0900
Subject: =?UTF-8?q?*=20src/base/ftobjs.c=20(Mac=5FRead=5FPOST=5FResource):?=
 =?UTF-8?q?=20Insert=20comments=20CVS-2014-9674-fixup-2=0Aand=20fold=20too?=
 =?UTF-8?q?=20long=20tracing=20messages.?=

(cherry picked from commit 1720e81e3ecc7c266e54fe40175cc39c47117bf5)
[benh: Backported to 2.4.2: don't use FT_THROW() or FT_ERR()]
---
 src/base/ftobjs.c | 34 ++++++++++++++++++++++++----------
 1 file changed, 24 insertions(+), 10 deletions(-)

--- a/src/base/ftobjs.c
+++ b/src/base/ftobjs.c
@@ -1545,21 +1545,28 @@
         goto Exit;
       if ( FT_READ_ULONG( temp ) )
         goto Exit;
-#if 0
-      FT_TRACE4(( "                 POST fragment #%d: length=0x%08x\n", i, temp));
+
+      /* FT2 allocator takes signed long buffer length,
+       * too large value causing overflow should be checked
+       */
+      FT_TRACE4(( "                 POST fragment #%d: length=0x%08x\n",
+                  i, temp));
       if ( 0x7FFFFFFFUL < temp )
       {
         error = FT_Err_Invalid_Offset;
         goto Exit;
       }
-#endif
 
       pfb_len += temp + 6;
     }
 
-    FT_TRACE2(( "             total buffer size to concatenate %d POST fragments: 0x%08x\n",
+    FT_TRACE2(( "             total buffer size to concatenate %d"
+                " POST fragments: 0x%08x\n",
                  resource_cnt, pfb_len + 2));
     if ( pfb_len + 2 < 6 ) {
+      FT_TRACE2(( "             too long fragment length makes"
+                  " pfb_len confused: 0x%08x\n",
+                  pfb_len ));
       error = FT_Err_Array_Too_Large;
       goto Exit;
     }
@@ -1584,13 +1591,16 @@
         goto Exit2;
       if ( FT_READ_ULONG( rlen ) )
         goto Exit2;
-#if 0
+
+      /* FT2 allocator takes signed long buffer length,
+       * too large fragment length causing overflow should be checked
+       */
       if ( 0x7FFFFFFFUL < rlen )
       {
         error = FT_Err_Invalid_Offset;
         goto Exit2;
       }
-#endif
+
       if ( FT_READ_USHORT( flags ) )
         goto Exit2;
       FT_TRACE3(( "POST fragment[%d]: offsets=0x%08x, rlen=0x%08x, flags=0x%04x\n",
@@ -1615,7 +1625,8 @@
         len += rlen;
       else
       {
-        FT_TRACE3(( "    Write POST fragment #%d header (4-byte) to buffer 0x%p + 0x%08x\n", i, pfb_data, pfb_lenpos ));
+        FT_TRACE3(( "    Write POST fragment #%d header (4-byte) to buffer"
+                    " 0x%p + 0x%08x\n", i, pfb_data, pfb_lenpos ));
         if ( pfb_lenpos + 3 > pfb_len + 2 )
           goto Exit2;
         pfb_data[pfb_lenpos    ] = (FT_Byte)( len );
@@ -1626,7 +1637,8 @@
         if ( ( flags >> 8 ) == 5 )      /* End of font mark */
           break;
 
-        FT_TRACE3(( "    Write POST fragment #%d header (6-byte) to buffer 0x%p + 0x%08x\n", i, pfb_data, pfb_pos ));
+        FT_TRACE3(( "    Write POST fragment #%d header (6-byte) to buffer"
+                    " 0x%p + 0x%08x\n", i, pfb_data, pfb_pos ));
         if ( pfb_pos + 6 > pfb_len + 2 )
           goto Exit2;
         pfb_data[pfb_pos++] = 0x80;
@@ -1645,7 +1657,8 @@
       if ( pfb_pos > pfb_len || pfb_pos + rlen > pfb_len )
         goto Exit2;
 
-      FT_TRACE3(( "    Load POST fragment #%d (%d byte) to buffer 0x%p + 0x%08x\n", i, rlen, pfb_data, pfb_pos ));
+      FT_TRACE3(( "    Load POST fragment #%d (%d byte) to buffer"
+                  " 0x%p + 0x%08x\n", i, rlen, pfb_data, pfb_pos ));
       error = FT_Stream_Read( stream, (FT_Byte *)pfb_data + pfb_pos, rlen );
       if ( error )
         goto Exit2;
@@ -1674,7 +1687,8 @@
 
   Exit2:
     if ( error == FT_Err_Array_Too_Large )
-      FT_TRACE2(( "  Abort due to too-short buffer to store all POST fragments\n" ));
+      FT_TRACE2(( "  Abort due to too-short buffer to store"
+                  " all POST fragments\n" ));
     else if ( error == FT_Err_Invalid_Offset )
       FT_TRACE2(( "  Abort due to invalid offset in a POST fragment\n" ));
     if ( error )
